Invoices.elm

```diff
@@ -24,45 +25,50 @@ type alias Product =
     , unit : Int
     , tax : Float
     , price : Float
-    , total : Float
     }
 
 
 type alias Model =
-    { total : Float
-    , products : List Product
+    { products : List Product
     }
 
 
 initModel : Model
 initModel =
-    Model 0 []
+    Model []
 
 
-createProduct : Int -> Product
-createProduct id =
-    Product id "" 1 21 0.0 0.0
+addNew : Model -> Model
+addNew { products } =
+    products
+        ++ [ { id = List.length products, item = "", unit = 1, tax = 21, price = 0 } ]
+        |> Model
 
 
 calculateProductTotal : Product -> Float
 calculateProductTotal product =
-    (calculateProductPriceTotal product) + (calculateProductTax product)
+    calculateProductPriceTotal product + calculateProductTax product
 
 
 calculateProductPriceTotal : Product -> Float
-calculateProductPriceTotal product =
-    let
-        unit =
-            toFloat product.unit
-    in
-        unit * product.price
+calculateProductPriceTotal { unit, price } =
+    toFloat unit * price
 
 
 calculateProductTax : Product -> Float
-calculateProductTax product =
-    (calculateProductPriceTotal product) * (product.tax / 100)
+calculateProductTax ({ tax } as product) =
+    calculateProductPriceTotal product * (tax / 100)
+
+
+updateWithId : id -> ({ a | id : id } -> { a | id : id }) -> List { a | id : id } -> List { a | id : id }
+updateWithId id predicate =
```

Comment:
this function is a bit odd since it's neither specific nor generall enough in my opinion but I wanted to demonstrate two things:

- some type polymorphism on records
- point-free style*

*as you can see this function takes 3 arguments but only 2 are declared. Also, There is not list supplied one line lower. This means that this declaration itself returns a function which takes a list as an argument and returns a new list.